### PR TITLE
Add Stackdriver Exporter for Opencensus

### PR DESCRIPTION
**Pull request by aLekSer**
Add Stackdriver exporter functionality into Opencensus metrics collecting system. Add agones.metrics.stackdriver helm config variable as well as change reporting period to 1 minute.
Now we can use Stackdriver or Prometheus to collect and monitor Agones stats.
Resulting metrics would have an agones/ prefix. For #144

### CODE REVIEW

**Comment by markmandel** [comment]
Quick thought - should there be a Gopkg.toml entry, so we are locking the new dependencies to a specific version?

**Comment by markmandel** [comment]
Looks like a lock timeout while the other test was running. Restarting.

**Comment by aLekSer** [comment]
Added Gopkg.toml.

File: install/helm/agones/README.md [hunk 1]
```diff
@@ -86,8 +86,9 @@ The following tables lists the configurable parameters of the Agones chart and t
 | `agones.rbacEnabled`                                | Creates RBAC resources. Must be set for any cluster configured with RBAC                        | `true`                 |
 | `agones.crds.install`                               | Install the CRDs with this chart. Useful to disable if you want to subchart (since crd-install hook is broken), so you can copy the CRDs into your own chart. | `true` |
 | `agones.crds.cleanupOnDelete`                       | Run the pre-delete hook to delete all GameServers and their backing Pods when deleting the helm chart, so that all CRDs can be removed on chart deletion | `true`          |  
-| `agones.metrics.enabled`                            | Enables controller metrics on port `8080` and path `/metrics`                                   | `true`                 |
+| `agones.metrics.prometheusEnabled`                            | Enables controller metrics on port `8080` and path `/metrics`                                   | `true`                 |
```
Comment (Kuqd) [inline install/helm/agones/README.md]: again I think we should have the same kind of name for prometheus and stackdriver.

File: install/helm/agones/templates/controller.yaml [hunk 2]
```diff
@@ -81,8 +81,10 @@ spec:
           value: {{ .Values.agones.image.sdk.alwaysPull | quote }}
         - name: SIDECAR_CPU_REQUEST
           value: {{ .Values.agones.image.sdk.cpuRequest | quote }}
-        - name: METRICS
-          value: {{ .Values.agones.metrics.enabled | quote }}
+        - name: PROMETHEUS
+          value: {{ .Values.agones.metrics.prometheusEnabled | quote }}
+        - name: STACKDRIVER_MONITORING
```
Comment (Kuqd) [inline install/helm/agones/templates/controller.yaml]: This env var name doesn't match with the doc.

File: pkg/metrics/exporter.go [hunk 3]
```diff
@@ -41,7 +42,42 @@ func RegisterPrometheusExporter(registry *prom.Registry) (http.Handler, error) {
...
+	sd, err := stackdriver.NewExporter(stackdriver.Options{
+		ProjectID: "",
+		// MetricPrefix helps uniquely identify your metrics.
+		MetricPrefix: "agones",
+	})
+	if err != nil {
+		return err
+	}
+	// It is imperative to invoke flush before your main function exits
+	defer sd.Flush()
```
Comment (Kuqd) [inline pkg/metrics/exporter.go]: this defer should be part of the main function ! here it will be triggered right after the registration call but not before the main exist.

File: cmd/controller/main.go [hunk 4]
```diff
@@ -181,7 +201,8 @@ func parseEnvFlags() config {
 	pflag.String(keyFileFlag, viper.GetString(keyFileFlag), "Optional. Path to the key file")
 	pflag.String(certFileFlag, viper.GetString(certFileFlag), "Optional. Path to the crt file")
 	pflag.String(kubeconfigFlag, viper.GetString(kubeconfigFlag), "Optional. kubeconfig to run the controller out of the cluster. Only use it for debugging as webhook won't works.")
-	pflag.Bool(enableMetricsFlag, viper.GetBool(enableMetricsFlag), "Flag to activate metrics of Agones. Can also use METRICS env variable.")
+	pflag.Bool(enablePrometheusMetricsFlag, viper.GetBool(enablePrometheusMetricsFlag), "Flag to activate metrics of Agones. Can also use PROMETHEUS env variable.")
+	pflag.Bool(enableStackdriverMetricsFlag, viper.GetBool(enableStackdriverMetricsFlag), "Flag to activate stackdriver monitoring metrics for Agones. Can also use STACKDRIVER env variable.")
```
[the remark this answers is not in the document]
Comment (aLekSer) [reply cmd/controller/main.go]: STACKDRIVER_MONITORING or STACKDRIVER_METRICS I think would be better

[the remark this answers is not in the document]
Comment (aLekSer) [reply cmd/controller/main.go]: Yes, sure, updated this

**Comment by aLekSer** [comment]
Adding Project ID to configuration yaml, could be useful to have

File: docs/metrics.md [hunk 5]
```diff
@@ -164,6 +168,45 @@ Open a web browser to [http://127.0.0.1:3000](http://127.0.0.1:3000), you should
...
+  metrics:
+    prometheusEnabled: false
+    prometheusServiceDiscovery: false
+    stackdriverEnabled: true
+    stackdriverProjectID: ""
+```
+
+With this configuration only Stackdriver exporter would be used instead of Prometheus exporter.
+
+Run the following command to install Agones to a cluster with an updated configuration:
```
Comment (Kuqd) [inline docs/metrics.md]: nitpick; I don't think we need to gives to command just ask to reinstall using helm.

File: docs/metrics.md [hunk 6]
```diff
@@ -164,6 +168,45 @@ Open a web browser to [http://127.0.0.1:3000](http://127.0.0.1:3000), you should
...
+```
+make install
```
Comment (markmandel) [inline docs/metrics.md]: Let's not put make commands in out user documentation if we can avoid it. Let's have explicit commands n our documentation instead.

File: install/helm/agones/README.md [hunk 7]
```diff
@@ -86,8 +86,10 @@ The following tables lists the configurable parameters of the Agones chart and t
...
-| `agones.metrics.enabled`                            | Enables controller metrics on port `8080` and path `/metrics`                                   | `true`                 |
+| `agones.metrics.prometheusEnabled`                  | Enables controller metrics on port `8080` and path `/metrics`                                   | `true`                 |
 | `agones.metrics.prometheusServiceDiscovery`         | Adds annotations for Prometheus ServiceDiscovery (and also Strackdriver)                        | `true`                 |
+| `agones.metrics.stackdriverEnabled`                 | (⚠️ **development feature**⚠️) Enables Stackdriver exporter of controller metrics              | `false`                 |
+| `agones.metrics.stackdriverProjectID`               | (⚠️ **development feature**⚠️) Project ID where current GKE cluster is deployed, if empty string project id from Application Default Credentials would be used. Used for Stackdriver.            | ``                 |
```
Comment (Kuqd) [inline install/helm/agones/README.md]: I would just say that this overrides the default gcp project id for use with stackdriver

File: docs/metrics.md [hunk 8]
```diff
@@ -164,6 +168,45 @@ Open a web browser to [http://127.0.0.1:3000](http://127.0.0.1:3000), you should
 
 > Makefile targets `make grafana-portforward`,`make kind-grafana-portforward` and `make minikube-grafana-portforward`.
 
+### Stackdriver installation
```
Comment (markmandel) [inline docs/metrics.md]: This will need to get moved over to the new website docs :cry:  sorry - someone was going to get bit by this.

File: docs/metrics.md [hunk 9]
```diff
@@ -164,6 +168,45 @@ Open a web browser to [http://127.0.0.1:3000](http://127.0.0.1:3000), you should
...
+
+Example of the dashboard appearance is provided below:
+
+![](stackdriver-metrics-dashboard.png)
+
+Currently there exists only manual way of configuring Stackdriver Dashboard. So it is up to you to set Alignment Period (minimal is 1 minute), group by parameter where you choose Metric label, filter parameter, etc.
+
+#### Troubleshooting
+If you could not see agones metrics please check the controller logs for connection errors. Check also that Stackdriver Monitoring API is enabled and user has all necessary permissions to access Stackdriver Monitoring, configure `stackdriverProjectID` manually, if default is not working.
```
Comment (Kuqd) [inline docs/metrics.md]:
```suggestion

If you can't see Agones metrics you should have a look at the controller logs for connection errors. Also ensure that your cluster has  the necessary credentials to interact with Stackdriver Monitoring. You can configure `stackdriverProjectID` manually, if the automatic discovery is not working.
```

File: docs/metrics.md [hunk 10]
```diff
@@ -50,7 +51,10 @@ Finally include that `ServiceMonitor` in your [Prometheus instance CRD](https://
 
 ### Stackdriver
 
-We don't yet support the [OpenCensus Stackdriver exporter](https://opencensus.io/exporters/supported-exporters/go/stackdriver/) but you can still use the Prometheus Stackdriver integration by following these [instructions](https://cloud.google.com/monitoring/kubernetes-engine/prometheus).
+We support the [OpenCensus Stackdriver exporter](https://opencensus.io/exporters/supported-exporters/go/stackdriver/). 
+In order to configure it you should enable Stackdriver Monitoring API in Google Cloud Console.
```
Comment (markmandel) [inline docs/metrics.md]: Let's link to the API here. Make life easy for people

File: docs/metrics.md [hunk 11]
```diff
@@ -50,7 +51,10 @@ Finally include that `ServiceMonitor` in your [Prometheus instance CRD](https://
...
+
+Also you can use the Prometheus Stackdriver integration by following these [instructions](https://cloud.google.com/monitoring/kubernetes-engine/prometheus).
```
Comment (markmandel) [inline docs/metrics.md]: Does it make sense to push prom+stackdriver at the same time we are saying use the stackdriver integration directly?

Is there any value add in using both layers together? if not, let's cut this piece I think.

File: docs/metrics.md [hunk 12]
```diff
@@ -164,6 +168,45 @@ Open a web browser to [http://127.0.0.1:3000](http://127.0.0.1:3000), you should
...
+
+Example of the dashboard appearance is provided below:
+
+![](stackdriver-metrics-dashboard.png)
+
+Currently there exists only manual way of configuring Stackdriver Dashboard. So it is up to you to set Alignment Period (minimal is 1 minute), group by parameter where you choose Metric label, filter parameter, etc.
+
+#### Troubleshooting
+If you could not see agones metrics please check the controller logs for connection errors. Check also that Stackdriver Monitoring API is enabled and user has all necessary permissions to access Stackdriver Monitoring, configure `stackdriverProjectID` manually, if default is not working.
+For example if wrong ProjectID is configured you would see such an error, same error will occure with Permissions is not configured in GCP:
```
Comment (Kuqd) [inline docs/metrics.md]:
```suggestion

Permissions problem example from controller logs:
```

File: docs/metrics.md [hunk 13]
```diff
@@ -164,6 +168,45 @@ Open a web browser to [http://127.0.0.1:3000](http://127.0.0.1:3000), you should
...
+
+Create a Fleet or a Gameserver in order to check that connection with stackdriver API is configured properly and so that you will be able to see the metrics data.
+
+Visit [Stackdriver monitoring](https://app.google.stackdriver.com) website, select your project, or choose `Create a new Workspace` and select GCP project where your cluster resides. In [Stackdriver metrics explorer](https://cloud.google.com/monitoring/charts/metrics-explorer) you should be able to find new metrics with prefix `agones/` (resource type is `Global`) after a couple of minutes. Choose the metrics you are interested in and add to a single or separate graphs. You can create multiple graphs, save them into your dashboard and use various aggregation parameters and reducers for each graph.
+
+Example of the dashboard appearance is provided below:
+
+![](stackdriver-metrics-dashboard.png)
+
+Currently there exists only manual way of configuring Stackdriver Dashboard. So it is up to you to set Alignment Period (minimal is 1 minute), group by parameter where you choose Metric label, filter parameter, etc.
```
[the remark this answers is not in the document]
Comment (aLekSer) [reply docs/metrics.md]: I reworded this paragraph a bit.

File: docs/metrics.md [hunk 14]
```diff
@@ -164,6 +168,45 @@ Open a web browser to [http://127.0.0.1:3000](http://127.0.0.1:3000), you should
 
 > Makefile targets `make grafana-portforward`,`make kind-grafana-portforward` and `make minikube-grafana-portforward`.
 
+### Stackdriver installation
+
+In order to use [Stackdriver monitoring](https://app.google.stackdriver.com) you should enable Stackdriver Monitoring API on Google Cloud Console. You need to grant all necessary permissions to the users (see [Access Control Guide](https://cloud.google.com/monitoring/access-control)). Stackdriver exporter uses a strategy called Application Default Credentials (ADC) to find your application's credentials. Details could be found here [Setting Up Authentication for Server to Server Production Applications](https://cloud.google.com/docs/authentication/production).
```
[the remark this answers is not in the document]
Comment (aLekSer) [reply docs/metrics.md]: It is enabled by default but I will add this useful link

**Comment by aLekSer** [comment]
Updated the docs and links in order to be compliant with the new website format.

File: docs/metrics.md (hunk 11 above)
Comment (aLekSer) [reply docs/metrics.md]: The only reason for that is to add golang metrics, like goroutines count, however it is not necessary to configure both at the same time.

File: site/content/en/docs/Guides/metrics.md [hunk 15]
```diff
@@ -54,10 +55,11 @@ Finally include that `ServiceMonitor` in your [Prometheus instance CRD](https://
 
 ### Stackdriver
 
-We don't yet support the [OpenCensus Stackdriver exporter](https://opencensus.io/exporters/supported-exporters/go/stackdriver/)
-but you can still use the Prometheus Stackdriver integration by following these [instructions](https://cloud.google.com/monitoring/kubernetes-engine/prometheus).
-Annotations required by this integration can be activated by setting the `agones.metrics.prometheusServiceDiscovery` 
-to true (default) via the [helm chart value]({{< relref "../Installation/helm.md" >}}).
+We support the [OpenCensus Stackdriver exporter](https://opencensus.io/exporters/supported-exporters/go/stackdriver/).
```
Comment (markmandel) [inline site/content/en/docs/Guides/metrics.md]: You get to play with our new documentation tools! We have "shortcodes" which you can wrap content in, to hide and show previous versions. See: https://agones.dev/site/docs/contribute/ for details.

TL;DR -

The old version should we wrapped in:

```
{{% feature expiryVersion="0.8.0" %}}
We don't yet support the OpenCensus Stackdriver exporter...
{{\% /feature %}}
```

And the new documentation should be wrapped in:
```
{{% feature publishVersion="0.8.0" %}}
We support the OpenCensus Stackdriver exporter...
{{\% /feature %}}
```

The tooling will show/hide each section appropriately as we release new versions.

File: site/content/en/docs/Guides/metrics.md [hunk 16]
```diff
@@ -174,6 +176,48 @@ Open a web browser to [http://127.0.0.1:3000](http://127.0.0.1:3000), you should
 
 > Makefile targets `make grafana-portforward`,`make kind-grafana-portforward` and `make minikube-grafana-portforward`.
 
+### Stackdriver installation
+
+In order to use [Stackdriver monitoring](https://app.google.stackdriver.com) you should enable Stackdriver Monitoring API on Google Cloud Console. You need to grant all the necessary permissions to the users (see [Access Control Guide](https://cloud.google.com/monitoring/access-control)). Stackdriver exporter uses a strategy called Application Default Credentials (ADC) to find your application's credentials. Details could be found here [Setting Up Authentication for Server to Server Production Applications](https://cloud.google.com/docs/authentication/production).
```
Comment (markmandel) [inline site/content/en/docs/Guides/metrics.md]:
```suggestion
In order to use [Stackdriver monitoring](https://app.google.stackdriver.com) you should [enable Stackdriver Monitoring API](https://cloud.google.com/monitoring/api/enable-api) on Google Cloud Console. You need to grant all the necessary permissions to the users (see [Access Control Guide](https://cloud.google.com/monitoring/access-control)). Stackdriver exporter uses a strategy called Application Default Credentials (ADC) to find your application's credentials. Details could be found here [Setting Up Authentication for Server to Server Production Applications](https://cloud.google.com/docs/authentication/production).
```

File: site/content/en/docs/Guides/metrics.md [hunk 17]
```diff
@@ -174,6 +176,48 @@ Open a web browser to [http://127.0.0.1:3000](http://127.0.0.1:3000), you should
...
+
+Note that Stackdriver monitoring is enabled by default on GKE clusters, however you can follow this [guide](https://cloud.google.com/kubernetes-engine/docs/how-to/monitoring#enabling_stackdriver_monitoring) if it was disabled on your GKE cluster.
+
+Default metrics exporter is Prometheus. In order to change it to Stackdriver update following helm parameters in {{< ghlink href="/install/helm/agones/values.yaml" branch="master" >}}values file{{< /ghlink >}}:
```
Comment (markmandel) [inline site/content/en/docs/Guides/metrics.md]: @Kuqd is this the right way to tell people to do this?  Or should we be directing people to the command line arguments?

File: site/content/en/docs/Installation/helm.md [hunk 18]
```diff
@@ -91,8 +91,10 @@ The following tables lists the configurable parameters of the Agones chart and t
 | `agones.rbacEnabled`                                | Creates RBAC resources. Must be set for any cluster configured with RBAC                        | `true`                 |
 | `agones.crds.install`                               | Install the CRDs with this chart. Useful to disable if you want to subchart (since crd-install hook is broken), so you can copy the CRDs into your own chart. | `true` |
 | `agones.crds.cleanupOnDelete`                       | Run the pre-delete hook to delete all GameServers and their backing Pods when deleting the helm chart, so that all CRDs can be removed on chart deletion | `true`          |  
-| `agones.metrics.enabled`                            | Enables controller metrics on port `8080` and path `/metrics`                                   | `true`                 |
+| `agones.metrics.prometheusEnabled`                  | Enables controller metrics on port `8080` and path `/metrics`                                   | `true`                 |
```
Comment (markmandel) [inline site/content/en/docs/Installation/helm.md]: Let's leave the old part's in here (we may need to move it to a new table, and do the same as we did above with the `feature` shortcode, and then take the new values, and move them to the table below that is wrapped already in the `feature` shortcode for the next version :+1:

File: site/content/en/docs/Guides/metrics.md (hunk 17 above)
Comment (Kuqd) [reply site/content/en/docs/Guides/metrics.md]: Helm users should know how set their values, I would give hint for that and not for modifying our own default values files.

Comment (aLekSer) [reply site/content/en/docs/Guides/metrics.md]: @markmandel I need to update this, but lets wait Cyril

**Comment by aLekSer** [comment]
Added feature shortcode and helm upgrade instructions. In order to test all three exporter options enabled you could run this helm command with image.tag:
```
helm upgrade --install --wait --set agones.metrics.stackdriverEnabled=true --set agones.metrics.prometheusEnabled=true --set agones.metrics.prometheusServiceDiscovery=true  agones /go/src/agones.dev/agones/install/helm/agones/ --set agones.image.tag=0.8.0-a96c1db    
```

File: site/content/en/docs/Contribute/_index.md [hunk 19]
```diff
@@ -60,7 +60,7 @@ or to hide a section from 0.8.0 onward:
 
 ```markdown
 {{\% feature expiryVersion="0.8.0" %}}
-  This is my special content that she be hidden <= 0.8.0
+  This is my special content that will be hidden >= 0.8.0
```
Comment (markmandel) [inline site/content/en/docs/Contribute/_index.md]: Good catch :+1:

File: site/content/en/docs/Guides/metrics.md [hunk 20]
```diff
@@ -174,6 +182,39 @@ Open a web browser to [http://127.0.0.1:3000](http://127.0.0.1:3000), you should
...
+
+Note that Stackdriver monitoring is enabled by default on GKE clusters, however you can follow this [guide](https://cloud.google.com/kubernetes-engine/docs/how-to/monitoring#enabling_stackdriver_monitoring) if it was disabled on your GKE cluster.
+
+Default metrics exporter is Prometheus. In order to change it to Stackdriver upgrade Agones release using helm with next three chart parameters changed:
+```
+helm upgrade --install --wait --set agones.metrics.stackdriverEnabled=true --set agones.metrics.prometheusEnabled=false --set agones.metrics.prometheusServiceDiscovery=false  agones ../install/helm/agones/
```
Comment (markmandel) [inline site/content/en/docs/Guides/metrics.md]:
```suggestion
helm install --name=my-release --namespace=agones-system --set agones.metrics.stackdriverEnabled=true --set agones.metrics.prometheusEnabled=false --set agones.metrics.prometheusServiceDiscovery=false  agones/agones
```

@Kuqd to confirm, but I think that is better?

File: site/content/en/docs/Guides/metrics.md [hunk 21]
```diff
@@ -174,6 +182,39 @@ Open a web browser to [http://127.0.0.1:3000](http://127.0.0.1:3000), you should
...
 
+{{% feature publishVersion="0.8.0" %}}
+### Stackdriver installation
+
+In order to use [Stackdriver monitoring](https://app.google.stackdriver.com) you should [enable Stackdriver Monitoring API](https://cloud.google.com/monitoring/api/enable-api) on Google Cloud Console. You need to grant all the necessary permissions to the users (see [Access Control Guide](https://cloud.google.com/monitoring/access-control)). Stackdriver exporter uses a strategy called Application Default Credentials (ADC) to find your application's credentials. Details could be found here [Setting Up Authentication for Server to Server Production Applications](https://cloud.google.com/docs/authentication/production).
+
+Note that Stackdriver monitoring is enabled by default on GKE clusters, however you can follow this [guide](https://cloud.google.com/kubernetes-engine/docs/how-to/monitoring#enabling_stackdriver_monitoring) if it was disabled on your GKE cluster.
+
+Default metrics exporter is Prometheus. In order to change it to Stackdriver upgrade Agones release using helm with next three chart parameters changed:
```
Comment (markmandel) [inline site/content/en/docs/Guides/metrics.md]:
```suggestion
Default metrics exporter is Prometheus. In order to install or upgrade Agones to use Stackdriver, use the following chart parameters:
```

File: site/content/en/docs/Guides/metrics.md (hunk 20 above)
Comment (Kuqd) [reply site/content/en/docs/Guides/metrics.md]: yes but you need agones before the release name, also does that works without adding our repo ? I'm wondering if helm hub expose our chart in the default repository.

Comment (markmandel) [reply site/content/en/docs/Guides/metrics.md]: So:
```
helm upgrade --install --wait --set agones.metrics.stackdriverEnabled=true --set agones.metrics.prometheusEnabled=false --set agones.metrics.prometheusServiceDiscovery=false agones/agones my-release-name 
```
That's better?

Comment (markmandel) [reply site/content/en/docs/Guides/metrics.md]: Also, helm hub doesn't add anything to the default repo - it's lists the installation instructions as:
`helm repo add agones https://agones.dev/chart/stable`
